src/data/nat/factorization.lean

```diff
@@ -0,0 +1,150 @@
+/-
+Copyright (c) 2021 Stuart Presnell. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Stuart Presnell
+-/
+import data.nat.prime
+import data.nat.mul_ind
+
+/-!
+# Prime factorizations
+
+ `n.factorization` is the finitely supported function `ℕ →₀ ℕ`
+ mapping each prime factor of `n` to its multiplicity in `n`.  For example, since 2000 = 2^4 * 5^3,
+  * `factorization 2000 2` is 4
+  * `factorization 2000 5` is 3
+  * `factorization 2000 k` is 0 for all other `k : ℕ`.
+-/
+
+open nat finset list finsupp
+open_locale big_operators
+
+namespace nat
+
+/-- `n.factorization` is the finitely supported function `ℕ →₀ ℕ`
+ mapping each prime factor of `n` to its multiplicity in `n`. -/
+noncomputable def factorization (n : ℕ) : ℕ →₀ ℕ := (n.factors : multiset ℕ).to_finsupp
+
+lemma factorization_eq_count {n p : ℕ} : n.factorization p = list.count p n.factors :=
```

Comment:
```suggestion
lemma factorization_eq_count {n p : ℕ} : n.factorization p = n.factors.count p :=
```
Do you think this should be a simp lemma in the opposite direction? As discussed on zulip, we should have exactly one canonical way of talking about this function.

Reply:
I've added a TODO note suggesting that this should be considered in the unification